lib/Exporter/ONNXModelWriter.cpp

```diff
@@ -0,0 +1,1107 @@
+    float value = SN->getValue();
+    if (value == 0.5f) {
+      proto->set_op_type("Sqrt");
+    } else if (value == -1.0f) {
+      proto->set_op_type("Reciprocal");
+    } else {
+      RETURN_ERR_IF_NOT(false, "Splat Node Value is invalid.");
+    }
+    break;
+  }
+  default:
+    proto->add_input(RHSN->getName());
+    break;
+  }
+
+  reporter(RHSN);
+  return llvm::Error::success();
+}
+
+llvm::Error ONNXModelWriter::writeTopK(const Node *node, GraphType &graph,
+                                       ReporterType &reporter) {
+  auto *proto = graph.add_node();
+  // Find dictionary entires
+  const auto *TKN = llvm::cast<TopKNode>(node);
+
+  addValueAttribute(proto, "k", TKN->getK());
+
+  return writeAllWithNode("topK", node, proto);
+}
+
+llvm::Error ONNXModelWriter::writePRelu(const Node *node, GraphType &graph,
+                                        ReporterType &reporter) {
+  auto *proto = graph.add_node();
+  proto->set_name(node->getName());
+  proto->set_op_type("PRelu");
+  proto->add_input(node->getNthInput(PReluNode::InputIdx).getNode()->getName());
+
+  const auto *slope = unwindBroadcastInput(
+      node->getNthInput(PReluNode::SlopeIdx).getNode(), reporter);
+  switch (slope->getKind()) {
+  case Kinded::Kind::SplatNodeKind: {
+    // Conversion a scalar to a tensor is required
+    Tensor scalar = {llvm::cast<SplatNode>(slope)->getValue()};
+    auto *tensorProto = graph.add_initializer();
+    tensorProto->set_name(slope->getName());
+    writeTensor(scalar, tensorProto);
+    proto->add_input(slope->getName());
+    reporter(slope);
+    break;
+  }
+  case Kinded::Kind::ReshapeNodeKind:
+    proto->add_input(
+        slope->getNthInput(ReshapeNode::InputIdx).getNode()->getName());
+    reporter(slope);
+    break;
+  default:
+    RETURN_ERR_IF_NOT(false,
+                      "Can't find Splat/Reshape Node as part of PRelu Node.");
+  }
+
+  outputsToProto(node, proto);
+  return llvm::Error::success();
+}
+
+llvm::Error ONNXModelWriter::writeGather(const Node *node, GraphType &graph,
+                                         ReporterType &reporter) {
+  auto *proto = graph.add_node();
+  // Find dictionary entires
+  const auto *GN = llvm::cast<GatherNode>(node);
+
+  auto batchDims = GN->getBatchDims();
+
+  if (batchDims != 0) {
+    addValueAttribute(proto, "axis", batchDims);
+    return writeAllWithNode("BatchGather", node, proto);
+  } else {
+    return writeAllWithNode("Gather", node, proto);
+  }
+}
+
+llvm::Error ONNXModelWriter::writeMatMul(const Node *node, GraphType &graph,
+                                         ReporterType &reporter) {
+  const bool isBatched = node->getKind() == Kinded::Kind::BatchMatMulNodeKind;
+  auto *proto = graph.add_node();
+  proto->set_name(node->getName());
+  proto->set_op_type("MatMul");
+
+  // Check if LHS/RHS are Transpose node
+  Node *LHS =
+      llvm::cast<TransposeNode>(node->getNthInput(BatchMatMulNode::LHSIdx));
+  if (LHS) {
```

Comment:
This is broken -- you need to use `llvm::dyn_cast<TransposeNode>` here, as `llvm::cast` does not return a nullptr if the type is incorrect.